src/main/kotlin/me/zeroeightsix/kami/util/MessageDetectionHelper.kt

```diff
@@ -0,0 +1,61 @@
+package me.zeroeightsix.kami.util
+
+import java.util.regex.Pattern
+
+object MessageDetectionHelper {
+    fun String.detectAndRemove(regex: Regexes): String? {
+        if (!this.detect(regex)) return null
+
+        val removed = regex.regex.replace(this, "")
+        return if (removed.isEmpty()) null else {
+            removed
+        }
+    }
+
+    fun String.detect(vararg regexes: Regexes): Boolean {
+        regexes.forEach {
+            if (it.regex.containsMatchIn(this)) return true
+        }
+        return false
+    }
+
+    fun String.detect(setting: Boolean, vararg regex: Regexes) = setting && this.detect(*regex)
+
+    fun isDirectReceived(direct: Boolean, message: String) =
+        message.detect(direct, Regexes.DIRECT, Regexes.DIRECT_ALT_1, Regexes.DIRECT_ALT_2)
+
+    fun isDirect(direct: Boolean, message: String) =
+        message.detect(direct, Regexes.DIRECT_SENT) || isDirectReceived(direct, message)
+
+    fun getDirectUsername(message: String): String? {
+        if (!isDirect(true, message)) return null
+
+        /* side note: this won't work if some glitched account has spaces in their username, but in all honesty, like 3 people globally have those */
```

Comment:
You're forgetting about cracked servers.

Reply:
How else could I make this work then? 

If someone is playing one of those servers they can obviously choose their own username, maybe display a warning to the user that it's only compatible without spaces? Feels hacky but I'm not sure how else

Reply:
Just match anything lazily, making sure that the 'username end marker' is matched explicitly. E.g. the vanilla message regex could look like this:
```regex
<(?<username>.*?)> (?<message>.*)
```
which would match
```
<my name has spaces> I'm inconsiderate.
```
just fine.

If a user has `>` in their name, that's an unfortunate edge case we won't try to cover. Also make sure that your code (and regex) does a sanity check, so that a horribly formatted chat message like:
```
<[086] cool boys in the house>
```
won't match a empty message with username `[086] cool boys in the house`.